src/v/cluster/archival/archiver_operations_impl.cc

```diff
@@ -171,23 +224,305 @@ class archiver_operations_impl : public archiver_operations_api {
+                    vlog(
+                      _rtclog.warn,
+                      "Only {} segments can't be admitted, segments: {}",
+                      num_accepted,
+                      metadata);
+                    metadata.resize(num_accepted);
+                }
+                // TODO: probe->gap_detected(...)
+                is_validated = true;
+            }
+            auto num_succeeded = metadata.size();
+            auto num_failed = upl_res.results.size() - num_succeeded;
+            // replicate metadata
+            // add clean command if needed
+            auto offset_to_opt = [](model::offset o) {
+                std::optional<model::offset> r;
+                if (o != model::offset{}) {
+                    r = o;
+                }
+                return r;
+            };
+
+            auto replication_result = co_await part->add_segments(
+              std::move(metadata),
+              offset_to_opt(upl_res.manifest_clean_offset),
+              offset_to_opt(upl_res.read_write_fence),
+              part->get_highest_producer_id(),
+              workflow_rtc.get_deadline(),
+              _as,
+              is_validated);
+
+            if (replication_result.has_error()) {
+                vlog(
+                  _rtclog.error,
+                  "Failed to replicate archival metadata: {}",
+                  replication_result.error());
+            }
+            co_return admit_uploads_result{
+              .ntp = upl_res.ntp,
+              .num_succeeded = num_succeeded,
+              .num_failed = num_failed,
+              .manifest_dirty_offset = replication_result.value(),
+            };
+        } catch (...) {
+            co_return error_outcome::unexpected_failure;
+        }
+        __builtin_unreachable();
     }
 
     /// Reupload manifest and replicate configuration batch
-    ss::future<result<manifest_upload_result>>
-    upload_manifest(retry_chain_node&, model::ntp) noexcept override {
-        co_return error_outcome::unexpected_failure;
+    ss::future<result<manifest_upload_result>> upload_manifest(
+      retry_chain_node& workflow_rtc, model::ntp ntp) noexcept override {
+        auto gate = _gate.hold();
+        try {
+            auto partition = _pm->get_partition(ntp);
+            const auto& manifest = partition->manifest();
+            const auto estimated_size = manifest.estimate_serialized_size();
+            auto key = partition->get_remote_manifest_path(manifest);
+            auto res = co_await _api->upload_manifest(
+              _bucket, manifest, std::move(key), workflow_rtc);
+            switch (res) {
+            case upload_result::cancelled:
+                co_return error_outcome::shutting_down;
+            case upload_result::timedout:
+                co_return error_outcome::timed_out;
+            case upload_result::failed:
+                co_return error_outcome::unexpected_failure;
+            default:
```

Comment:
nit: it might be safer to be explicit here and match on success, in case more variants are added later.